libslide/src/evaluator_rules/rule.rs

```diff
@@ -1,19 +1,59 @@
 use super::pattern_match::{MatchRule, PatternMatch};
 use crate::grammar::*;
-use crate::utils::hash;
+use crate::utils::{hash, indent, unique_pats};
 use crate::{parse_expression_pattern, scan};
 
 use core::fmt;
 use std::collections::HashMap;
+use std::error::Error;
 use std::rc::Rc;
 
 /// A mapping between two expression patterns.
-#[derive(Clone)]
+#[derive(Clone, Debug)]
 pub struct PatternMap {
     pub from: Rc<ExprPat>,
     pub to: Rc<ExprPat>,
 }
 
+impl fmt::Display for PatternMap {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{} -> {}", self.from.to_string(), self.to.to_string())
+    }
+}
+
+#[derive(Debug)]
+pub struct UnresolvedMapping {
+    map: PatternMap,
+    unresolved_pats: Vec<Rc<ExprPat>>,
+}
+
+impl fmt::Display for UnresolvedMapping {
+    #[rustfmt::skip]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut missing_pats = self
+            .unresolved_pats
+            .iter()
+            .map(|p| format!(r#""{}""#, p))
+            .collect::<Vec<_>>();
+        missing_pats.sort();
+        let missing_pats = missing_pats.join(", ");
+
+        write!(
+            f,
+r#"Could not resolve pattern map
+{}"{}"
+Specifically, source "{}" is missing pattern(s) {} present in target "{}""#,
+            indent("\n", 4),
+            self.map,
+            self.map.from,
+            missing_pats,
+            self.map.to
+        )
+    }
+}
+
+impl Error for UnresolvedMapping {}
```

Comment:
why do we need this?

Reply:
Error is the generic error trait in Rust. This means we can pass it along as an Error in a function that accepts or returns an Error, without having to explicitly know it’s an UnresolvedMapping. Basically the reason you would use a trait (interface).